Risk Calculation does not work because DiagnosisKeysDataMapping is not applied (EXPOSUREAPP-3895)

kolyaopahle opened the pull request:
### Description
This PR includes config checking for the DiagnosisKeysDataMapping and applies it to the enfclient if configuration is deemed to have changed

kolyaopahle committed:
We now check for the DiagnosisKeyDataMapping and apply it to the ENFClient if changed

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigChangeDetector.kt:
```diff
@@ -60,4 +64,19 @@ class ConfigChangeDetector @Inject constructor(
+        val oldDiagnosisKeyDataMapping = enfClient.getDiagnosisKeysDataMapping()
+        val newDiagnosisKeyDataMapping = exposureWindowRiskCalculationConfig.diagnosisKeyDataMapping
+
+        if (oldDiagnosisKeyDataMapping != newDiagnosisKeyDataMapping) {
```
Does this compare object references or compare by value?

kolyaopahle replied:
should be an equal comparison, not 100% sure if its enough though, will do some checking

kolyaopahle replied:
Comparison has been changed and is now unittested

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigChangeDetector.kt:
```diff
@@ -60,4 +64,19 @@ class ConfigChangeDetector @Inject constructor(
     }
+
+    @VisibleForTesting
+    internal suspend fun checkForDiagnosisKeysDataMapping(exposureWindowRiskCalculationConfig: ExposureWindowRiskCalculationConfig) {
```
Is this also invoked if the app config cannot be loaded initially and the client falls back to the default app config?

kolyaopahle replied:
config change detector is no longer used

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigChangeDetector.kt:
```diff
@@ -60,4 +64,19 @@ class ConfigChangeDetector @Inject constructor(
+
+    @VisibleForTesting
+    internal suspend fun checkForDiagnosisKeysDataMapping(exposureWindowRiskCalculationConfig: ExposureWindowRiskCalculationConfig) {
+        val oldDiagnosisKeyDataMapping = enfClient.getDiagnosisKeysDataMapping()
```
This is the _current_ diagnosisKeysDataMapping, right? Might be worth adjusting the variable name accordingly 😉

kolyaopahle replied:
fixed in new place

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ExposureWindowRiskCalculationConfig.kt:
```diff
@@ -13,6 +14,7 @@ interface ExposureWindowRiskCalculationConfig {
         List<RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping>
     val normalizedTimePerDayToRiskLevelMappingList:
         List<RiskCalculationParametersOuterClass.NormalizedTimeToRiskLevelMapping>
+    val diagnosisKeyDataMapping: DiagnosisKeysDataMapping
```
Typo: key**s** (plural)

kolyaopahle replied:
fixed typo

mlenkeit commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/ConfigChangeDetector.kt:
```diff
@@ -60,4 +64,19 @@ class ConfigChangeDetector @Inject constructor(
+        if (oldDiagnosisKeyDataMapping != newDiagnosisKeyDataMapping) {
+            try {
+                Timber.i("New DiagnosisKeysDataMapping differs from last one, applying.")
+                enfClient.setDiagnosisKeysDataMapping(newDiagnosisKeyDataMapping)
```
This can throw exceptions for various reasons, at least:
1. rate limit error
1. invalid param

Only the rate limit error should be silent, other errors should not (otherwise, we wouldn't figure out that there's an error)

kolyaopahle replied:
new implementation rethrows error if status code is not Rate Limit

kolyaopahle committed:
Moved diagnosisKeysDataMapping to windowProvider

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Renaming and linting

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/ExposureWindowRiskCalculationConfigMapper.kt:
```diff
@@ -34,10 +41,24 @@ class ExposureWindowRiskCalculationConfigMapper @Inject constructor() :
         )
     }
 
+    private fun AppConfigAndroid.ApplicationConfigurationAndroid.diagnosisKeysDataMapping():
```
Why are we creating the mapping, and then taking the mapping, from protobuf class to protobuf class?
Could we use a data class here, to keep the protobuf classes hidden?

kolyaopahle replied:
the second class is not a protobuf but the internal enf config used by google, we could also implement a data class here, i oriented the initial implementation against the old mapRiskScoreToExposureConfiguration https://github.com/corona-warn-app/cwa-app-android/pull/1677/files#diff-0b01d7464a4e38df4e564db97b132516563e7b4ff1d004f4b6b0110d193f79edL67

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeysdatamapper/DefaultDiagnosisKeysDataMapper.kt:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    private fun DiagnosisKeysDataMapping.hasChanged(old: DiagnosisKeysDataMapping?): Boolean {
```
Please add unit-tests for this.
* Direct matching
* Empty matching
* Null matching
* whole map changed
* one key in a map changed
* one value in a map changed

kolyaopahle replied:
Unit tests have been added

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeysdatamapper/DefaultDiagnosisKeysDataMapper.kt:
```diff
@@ -0,0 +1,57 @@
+        }
+
+    override suspend fun updateDiagnosisKeysDataMapping() {
+        val currentDiagnosisKeysDataMapping = getDiagnosisKeysDataMapping()
```
Assume that `getDiagnosisKeysDataMapping` can fail too, try catch + log it and early abort.

kolyaopahle replied:
should now be caught and handled appropriately

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeysdatamapper/DefaultDiagnosisKeysDataMapper.kt:
```diff
@@ -0,0 +1,57 @@
+
+    override suspend fun updateDiagnosisKeysDataMapping() {
+        val currentDiagnosisKeysDataMapping = getDiagnosisKeysDataMapping()
+        val newDiagnosisKeysDataMapping = appConfigProvider
```
I think it's nicer if we pass `ExposureWindowRiskCalculationConfig` via function parameter. Updating the mapping is always tied to a submission, as the app config can change asynchronously, this could lead to 2 different configs being used in the task and here.

Task gets Appconfig -> ENFClient provide diagnosisKeys(config, files)

kolyaopahle replied:
Invocation has been moved to provide diagnosis key task

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeysdatamapper/DefaultDiagnosisKeysDataMapper.kt:
```diff
@@ -0,0 +1,57 @@
+
+        if (newDiagnosisKeysDataMapping.hasChanged(currentDiagnosisKeysDataMapping)) {
+            try {
+                Timber.i("New DiagnosisKeysDataMapping differs from last one, applying.")
```
Let's log the old and the new config if this happens.

kolyaopahle replied:
has been added

kolyaopahle committed:
Moved app config fetch to key provide task, added error handling, added logging to config replace

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Fixed old unit tests and implemented tests for config comparison,
TestForAPIFragment is currently broken

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Linting

kolyaopahle committed:
Added more unit testing for DefaultDiagnosisKeysDataMapper and fixed TestForAPIFragment

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

kolyaopahle committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

mlenkeit committed:
Update default app config to latest version

This includes a change for the default for reportTypeWhenMissing from
0 to 1, as 0 is not accepted by setDiagnosisKeysDataMapping.

https://github.com/corona-warn-app/cwa-app-tech-spec/pull/10/commits/090801921f4a8581543a9f716fd0572c12901c71
https://github.com/corona-warn-app/cwa-protocol-buffers/commit/38a641e2a808c41d225e9fd09d27ec7959551412

kolyaopahle committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

kolyaopahle committed:
Fixed Unittests

harambasicluka committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

kolyaopahle committed:
Merge remote-tracking branch 'remotes/origin/release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

Signed-off-by: Kolya Opahle <k.opahle@sap.com>

# Conflicts:
#	Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/api/ui/TestForAPIFragment.kt

kolyaopahle committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

BMItr reviewed (approved):
lgtm :ok_hand:

kolyaopahle committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

kolyaopahle committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping

ralfgehrer committed:
Merge branch 'release/1.8.x' into fix/3895-apply-diagnosis-key-data-mapping
